cirq/ops/raw_types.py

```diff
@@ -460,6 +468,104 @@ def validate_args(self, qubits: Sequence['cirq.Qid']):
         _validate_qid_shape(self, qubits)
 
 
+class Tag:
+    """A wrapper around string that can be used to differentiate operations.
+
+    This can be instantiated using a string value or sub-classed to
+    differentiate or tag specific instances of gates.
+
+    Tags can then be attached to an operation using Operation.with_tags(),
+    which will return a TaggedOperation or by instantiating a TaggedOperation
+    directly.
+
+    This can be used for special processing of certain operations.  For
+    instance, to prevent optimization or decomposing special gates.
+    """
+
+    def __init__(self, value: Optional[str] = None):
+        if value:
+            self.value = value
+        else:
+            self.value = ''
+
+    def __hash__(self):
+        return hash((
+            self.__class__,
+            self.value,
+        ))
+
+    def __eq__(self, other):
+        return (isinstance(other, self.__class__) and
+                isinstance(self, other.__class__) and self.value == other.value)
+
+    def __str__(self):
+        class_name = self.__class__.__name__
+        if class_name == 'Tag':
+            class_name = 'cirq.Tag'
+        return f"{class_name}('{self.value}')"
+
+    def __repr__(self):
+        return str(self)
+
+    def _json_dict_(self):
+        return protocols.obj_to_dict_helper(self, ['value'])
+
+
+@value.value_equality
+class TaggedOperation(Operation):
+    """A specific operation instance that has been identified with a set
+    of Tags for special processing.  This can be initialized with
+    Using Operation.with_tags(tag) or by TaggedOperation(op, tag).
+    """
+
+    def __init__(self, sub_operation: 'cirq.Operation', *tags: 'cirq.Tag'):
```

Comment:
I'd change the type of `tags` to just `Any`.

Reply:
Done.